design-documents/testing-strategy.md

```diff
@@ -0,0 +1,39 @@
+## Updated Magento Commerce Testing Strategy
+
+### Terms
+
+<!-- Describe any new terms used in the document -->
+
+### Overview
+
+<!-- Describe what we are trying to solve and a few sentence summary of design approach. -->
+
+Despite running 20,000 - 30,000 tests per pull request, we have little to no confidence in the stability of the product. This means that we still need a manual regression process that lasts weeks in the best of cases. 
+
+A sizable portion of this can be attributed to our current high-level testing mentality that prioritizes system-level black box testing over other types of tests. While these types of tests are necessary and good to have, they only provide implicit coverage to the code paths of the concrete implementations that happen to be loaded in the high-level black box tests (they are also expensive/difficult to maintain). We still need explicit coverage for the concrete implementations that extension developers have hard dependencies on. This is true for the simple code paths as well as variations both positive and negative flows.
+
+### Design
+
+#### New Testing Strategy:
+
+The new strategy for testing relies on having test coverage at each layer. The idea is to have many small and fast tests. Therefore, tests MUST be written in the following order of granularity and the higher-level the test, the fewer tests there will be:
+
+1. Code MUST be tested on its own. This MAY be done through unit or integration testing as appropriate. This level of test coverage ensures that the code can properly function on its own.  
+1. Code MUST be shown to work with known collaborators. This MUST be done through integration tests and ideally SHOULD involve as few components as possible per test case. This level of tests ensures that the code will work with its known associates. Examples of such associates would include aggregators, composites, etc.
+1. webapi behavior MUST be covered by blackbox tests to assert that the service contracts are properly satisfied. This ensures that the software package as a whole functions properly when everything is compiled together. This includes REST/SOAP/GraphQl etc. To be clear, these tests do not exclude the underlying concrete implementations from being tested. 
+1. MFTF tests SHOULD only be used to cover P0/P1 testing scenarios. Until these scenarios are all completed, a product owner MUST be consulted for approval of new tests. Additionally, MFTF MUST NOT be used for testing scenarios that are not P0/P1
```

Comment:
I believe this should be severity, not priority scenarios. Severity defines impact on product, priority is used to specify order in which issues that have the same priority should be fixed.

Reply:
> I believe this should use severity, not priority. 
and 
> coverage for S2 scenarios might be important for regression.

I will defer these to @irenedhong for this. She defined this policy. However I will comment that this is something that I'm personally conflicted with which direction we should go because while I understand Irene's policy I also see that we have separate policy for how we handle P2/S2 bugs compared to 3's and 4's and I'm not sure it's clear why we would have a separate policy for writing the tests versus fixing the bugs.

> One of the benefits of functional tests is that they are declarative and can be extended by extension developers. What is our strategy regarding other types of tests that are not declarative?

This is something that has been discussed in depth. It's mentioned in the proposal but this is only a theoretical problem currently. We understand this could potentially be something that needs to be addressed in the future but we have been unable to find any examples (even hypothetical) of when this was actually a problem and so we can't try to create some solution for it. Also, should this ever become a problem, we aren't eliminating functional tests so they could be used as a solution while some extensibility is built to accommodate these issues.

Reply:
@maghamed recommended that we use Priority over Severity because something can be low priority overall but high severity for a specific client. I leave this more up to those who have more knowledge between the differences and I have no opinion on this whether we choose to say it is P0/P1 or S0/S1.

For the distinction that we do this arbitrary cut off at P2 for UI tests was a recommendation by @maghamed that I agreed with. Please remember that we will continue to test P2 and are actually investing quite a lot of effort to shore these variations/edge cases/negative cases/permutations in our code base for the most critical modules. The note here is that we want to cover these with cheaper and faster non GUI tests as there can be hundreds of permutations/edge cases/negative/variations. We want to save the use of highly valuable but very expensive GUI tests for the most important P0/P1 scenarios where we are willing to pay the price of maintenance, creation, and run time. As part of this effort, we are spelling out what the most critical P0/P1 scenarios even are. Also we allow for exceptions to the rule by asking the product owner if we think it is critical to have a P2 covered by a GUI test. But then that begs to ask, if it is so critical why is it not a P0/P1?

For our bug fixing policy, I believe this document aligns with that. We hotfix for any P0 scenarios and we will stop a release for P1 scenarios. While we do fix P2's, we don't fix all P2's because the business value is not always there as they are often variations/permutations/edge cases/negative cases.

@melnikovi can you please clarify what you mean 'coverage for S2/P2 scenarios might be important for regression'. Which scenarios? Would you say that automated coverage for S2/P2 scenarios are not enough with the effort underway? Who would define all these P2 scenarios that need to be covered in regression specifically?